// ef-client/index.js
"use strict";
exports.__esModule = true;
var url = require("url");
var q = require("q");
var https = require("https");
var querystring = require("querystring");
var EFClient = (function () {
    function EFClient(endpoint, username, password, skipCertCheck) {
        this.endpoint = url.parse(endpoint + "/rest/v1.0");
        if (!this.endpoint.host) {
            throw new Error("No hostname found");
        }
        this.username = username;
        this.password = password;
        this.skipCertCheck = skipCertCheck;
    }
    EFClient.prototype.getProject = function (projectName) {
        var promise = this.get("/projects/" + querystring.escape(projectName), undefined);
        return promise;
    };
    EFClient.prototype.getPipeline = function (pipelineName, projectName) {
        var promise = this.get("/pipelines/" + querystring.escape(pipelineName), { projectName: projectName });
        return promise;
    };
    EFClient.prototype.runPipeline = function (pipelineName, projectName) {
        return this.post("/pipelines", { pipelineName: pipelineName, projectName: projectName }, "");
    };
    // runPipeline(pipelineName: string, projectName: string) {
    //     let projectPromise = this.getProject(projectName);
    //     projectPromise.then((res) => {
    //         return this.getPipeline(pipelineName, projectName);
    //     }).then((res) => {
    //         return this.post("/pipelines");
    //     }).catch((e) => {
    //         console.log(e);
    //     });
    //     // let runPipelinePromise = pipelinePromise.then((res) => {
    //     //     return this.post("/pipelines", {pipelineName: pipelineName, projectName: projectName}, "");
    //     // }).catch((e) => {
    //     //     console.log("Error occured: ", e);
    //     // });
    //     // return runPipelinePromise;
    // }
    EFClient.prototype.request = function (path, method, query, payload) {
        var def = q.defer();
        var endpoint = this.endpoint;
        var port = endpoint.port ? parseInt(endpoint.port) : 443;
        if (this.skipCertCheck) {
            process.env.NODE_TLS_REJECT_UNAUTHORIZED = "0";
        }
        var queryString = "";
        if (query) {
            var pairs = new Array();
            for (var key in query) {
                var value = query[key];
                pairs.push(key + "=" + querystring.escape(value));
            }
            queryString = '?' + pairs.join("&");
        }
        var options = {
            host: endpoint.hostname,
            port: port,
            method: method,
            path: endpoint.path + path + queryString,
            auth: this.username + ':' + this.password,
            headers: {
                'Accept': 'application/json',
                'Content-Type': 'application/json'
            }
        };
        console.log(options);
        var responseString = "";
        var req = https.request(options, function (res) {
            res.setEncoding('utf8');
            res.on('data', function (chunk) {
                responseString += chunk;
            });
            res.on('end', function () {
                var statusCode = res.statusCode;
                if (statusCode == 200) {
                    var responseObject = JSON.parse(responseString);
                    def.resolve(responseObject);
                }
                else {
                    def.reject({ statusCode: statusCode, response: JSON.parse(responseString) });
                }
            });
        }).on('error', function (e) {
            console.log('http request error');
            def.reject(e);
        });
        if (payload) {
            req.write(payload);
        }
        req.end();
        return def.promise;
    };
    EFClient.prototype.post = function (path, query, payload) {
        return this.request(path, 'POST', query, payload);
    };
    EFClient.prototype.put = function (path, query, payload) {
        return this.request(path, 'PUT', query, payload);
    };
    EFClient.prototype.get = function (path, query) {
        return this.request(path, 'GET', query, undefined);
    };
    EFClient.prototype.__get = function (path) {
        var def = q.defer();
        var endpoint = this.endpoint;
        var port = endpoint.port ? parseInt(endpoint.port) : 443;
        process.env.NODE_TLS_REJECT_UNAUTHORIZED = "0";
        var options = {
            host: endpoint.hostname,
            port: port,
            path: endpoint.path + path,
            auth: this.username + ':' + this.password,
            headers: {
                'Accept': 'application/json',
                'Content-Type': 'application/json'
            }
        };
        console.log('GET http://' + options.host + ':' + options.port + options.path);
        var responseString = "";
        https.get(options, function (res) {
            res.setEncoding('utf8');
            res.on('data', function (chunk) {
                responseString += chunk;
            });
            res.on('end', function () {
                var statusCode = res.statusCode;
                if (statusCode == 200) {
                    var responseObject = JSON.parse(responseString);
                    def.resolve(responseObject);
                }
                else {
                    def.reject({ statusCode: statusCode, response: JSON.parse(responseString) });
                }
            });
        }).on('error', function (e) {
            console.log('http request error');
            def.reject(e);
        });
        return def.promise;
    };
    return EFClient;
}());
exports.EFClient = EFClient;
;
;
